feat: integrate block number loading for date-to-block conversion

yahgw-ai opened the pull request:
## Summary

Implements block number loading and date-to-block conversion in RecipientRecievedScanner as specified in issue #178.

## Changes

- Added `convertDatesToBlockRanges()` method to convert date ranges to block number ranges
- Added `validateBlockNumbersForDates()` to ensure all required block numbers exist before processing
- Added `convertDateToBlockRange()` to convert individual dates to start/end block pairs
- Implemented proper calculation of start blocks (previous day's end block + 1, or block 1 for first date)
- Added comprehensive error handling with descriptive messages for missing block numbers

## Test Plan

Followed strict TDD approach:
- RED phase: Added comprehensive failing tests for all requirements
- GREEN phase: Implemented minimal code to make tests pass
- REFACTOR phase: Cleaned up implementation for clarity and maintainability

All new tests pass, validating:
- ✅ Date to block number conversion
- ✅ Error handling for missing block numbers
- ✅ Block range creation with start/end blocks
- ✅ Validation before processing
- ✅ Edge case handling (first date in block data)

## Notes

- This PR only handles the conversion logic as specified in the issue
- Actual event querying, chunking, and processing are out of scope (marked with TODO)
- Implementation is ready for the next phase of development

Closes #178

yahgw-ai committed:
test: add failing tests for block number conversion (RED phase)

Add comprehensive test suite for converting dates to block numbers
including validation, error handling, and edge cases.

yahgw-ai committed:
feat: implement block number conversion for date ranges (GREEN phase)

Add methods to convert dates to block numbers:
- validateBlockNumbersForDates: validates all required blocks exist
- convertDateToBlockRange: converts date to start/end block numbers
- Handles edge case for first date (starts from block 1)
- Throws descriptive errors for missing block numbers

yahgw-ai committed:
refactor: improve block number conversion implementation (REFACTOR phase)

- Combine validation and conversion into single method
- Extract getStartBlockForDate for clarity
- Improve code organization and readability
- Maintain backward compatibility for tests
- Add descriptive comments

yahgwai commented on src/recipient-recieved-scanner.ts:
```diff
@@ -81,7 +81,15 @@ export class RecipientRecievedScanner {
 
-      // TODO: Process the date range (out of scope for this issue)
+      // Convert dates to block ranges and validate block numbers
+      const blockRanges = this.convertDatesToBlockRanges(
```
We're overcomplicating this. We don't need to first calculate date ranges and then calculate block ranges from that. We can just directly calculate block ranges.

yahgwai reviewed (commented):
reoipening comment - unnapprove

yahgw-ai committed:
refactor: simplify block range calculation by combining date range and conversion steps

Per PR feedback, removed the intermediate date range to block range conversion step.
Now directly calculate block ranges within determineBlockRangesToProcess() method.
This simplifies the code flow while maintaining the same behavior.

yahgw-ai committed:
refactor: simplify to process distributors one day at a time

Implement cleaner approach suggested in PR review:
- Process each distributor day by day instead of calculating all block ranges upfront
- Remove intermediate getDateRange method
- Directly iterate through dates and process block ranges as we go
- Makes the flow more straightforward and easier to follow

yahgw-ai committed:
test: fix failing tests by adding missing block data and fake timers

- Add fake timers to 'scan - loading block numbers' and 'scan - loading existing event data' test suites
- Add block data for 2022-07-10 to support test scenarios
- Update test expectations to match the new block data structure
- Adjust 'throws error when block number is missing' test to use 2022-07-09
- Update 'handles first date in block data correctly' test to use 2022-07-10
